### ISSUE TITLE
Linking error with an @EnableReflectiveInstantiation object inside a class

### ISSUE THREAD

**Issue by sjrd** [scala-js/scala-js#3228]
The following code causes a linking error:
```scala
class Foo {
  @EnableReflectiveInstantiation
  object Bar
}
```
The error is:
```
[error] Cannot access module for non-module helloworld.Foo$Bar$
[error]   called from helloworld.Foo$Bar$.<clinit>()
[error]   called from core module analyzer
```
This happens because, as a nested `object`, `Bar` is not really a module class, but the codegen for `@EnableReflectiveInstantiation` believes `Bar$` is a module class, producing the following erroneous IR:
```scala
class Lhelloworld_Foo$Bar$ extends O {
  def init___Lhelloworld_Foo($$outer: Lhelloworld_Foo) {
    this.O::init___()
  }
  static def clinit___() {
    mod:sjs_reflect_Reflect$.registerLoadableModuleClass__T__jl_Class__sjs_js_Function0__V("helloworld.Foo$Bar$", classOf[Lhelloworld_Foo$Bar$], (lambda<>() = {
      mod:Lhelloworld_Foo$Bar$
    }))
  }
}
```

**Comment by cheeseng** [comment]
Hi @sjrd, @bvenners, 

Thanks for the workaround, the link problem seems gone but now when we're trying to run our tests it hits: 

```
[info] Fast optimizing /home/cheeseng/git/scalatest/scalatest-test.js/target/scala-2.11/scalatesttestjs-test-fastopt.js
[error] buffer.js:385
[error]     throw new Error('toString failed');
[error]     ^
[error] 
[error] Error: toString failed
[error]     at Buffer.toString (buffer.js:385:11)
[error]     at Object.fs.readFileSync (fs.js:461:33)
[error]     at Object.Module._extensions..js (module.js:416:20)
[error]     at Module.load (module.js:344:32)
[error]     at Function.Module._load (module.js:301:12)
[error]     at Module.require (module.js:354:17)
[error]     at require (internal/module.js:12:17)
[error]     at [stdin]:38:1
[error]     at Object.exports.runInThisContext (vm.js:54:17)
[error]     at Object.<anonymous> ([stdin]-wrapper:6:22)
org.scalajs.jsenv.ExternalJSEnv$NonZeroExitException: Node.js exited with code 1
	at org.scalajs.jsenv.ExternalJSEnv$AbstractExtRunner.waitForVM(ExternalJSEnv.scala:119)
	at org.scalajs.jsenv.ExternalJSEnv$ExtRunner.run(ExternalJSEnv.scala:168)
	at org.scalajs.sbtplugin.FrameworkDetector.detect(FrameworkDetector.scala:71)
	at org.scalajs.sbtplugin.ScalaJSPluginInternal$$anonfun$79.apply(ScalaJSPluginInternal.scala:1004)
	at org.scalajs.sbtplugin.ScalaJSPluginInternal$$anonfun$79.apply(ScalaJSPluginInternal.scala:981)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
[error] (scalatestTestJS/test:loadedTestFrameworks) org.scalajs.jsenv.ExternalJSEnv$NonZeroExitException: Node.js exited with code 1
[error] Total time: 91 s, completed Dec 19, 2017 2:00:42 PM
```

Any idea what is causing it?

Thanks!

**Comment by cheeseng** [comment]
@sjrd @bvenners Sorry forgot to mention that the buffer.js error happened in 2.11 but it builds and runs fine on 2.12 also.

**Comment by sjrd** [comment]
Would you mind opening another issue about this, as it seems totally unrelated to the one described in this issue?

**Comment by cheeseng** [comment]
@sjrd Sure, here's the new issue created: 

https://github.com/scala-js/scala-js/issues/3230

Thanks for your help!

cc @bvenners

**Comment by sjrd** [comment]
Fixed in e28ff5e0444c1410c2b44d4f819ea6b24966c570.
